### ISSUE TITLE
Jax' transfer guard and XLA-CPU

### ISSUE THREAD

**Issue by akolesnikoff** [google/jax#16602]
### Description

Jax' transfer guard  potentially has two problems related to the CPU-placed arrays.

1. Transfering numpy (or pure python) arrays/scalars to XLA-CPU device is considered to be host2device transfer:
```
jax.jit(jax.random.PRNGKey, backend='cpu')(np.array(0))
```
Transfer guard says: `host-to-device transfer: aval=ShapedArray(int32[]), dst_sharding=GSPMDSharding({replicated})`

Ideally I want to be able to ignore `numpy -> XLA-CPU` transfers and only trigger the transfer guard for transfers to the accelerators.

2. Interestingly, XLA-CPU -> numpy is not triggering transfer guard:
```
np.array(jax.device_put(0, device=jax.devices('cpu')[0]))
```
Transfer guard says: 🤐.

### Which accelerator(s) are you using?

TPU

**Comment by jakevdp** [comment]
Thanks for the report!

I suspect the reason for this is that while XLA:CPU → NumPy transfers can always be done in a zero-copy fashion, NumPy → XLA:CPU transfers cannot always be done in this manner. The reason for this asymmetry is that XLA arrays have stricter requirements regarding byte layout than do NumPy arrays.

@hyeontaek may be able to provide more context on the behavior you observe here

**Comment by hyeontaek** [comment]
Yes, as @jakevdp pointed out, the asymmetry came from the zero-copy vs. copy difference in two transfer directions.

**Comment by akolesnikoff** [comment]
I see, thanks for the clarification regarding the asymmetry.

It was actually not my main question, sorry for obscure phrasing.

The main problem is that I can not configure transfer guard to disallow implicit accelerator transfers, while allowing XLA:CPU transfers. Normally, I do not care much about host/cpu memory, but care about accelerator-related transfers and memory. Do you think you can add something like `jax.transfer_guard("disallow_accelerator)` or  `jax.transfer_guard("disallow_tpu)`?

**Comment by hyeontaek** [comment]
There seem to be a few potential high-level design options:

1. Focusing on logical operations that involve converting NumPy array and JAX array, or transferring JAX arrays to different devices: This is semantically clean/succinct though it may not necessarily be the not useful form of the API. In this direction, we will want to make XLA:CPU to NumPy to be guarded and remove the asymmetry.

2. Focusing on involved backend or platform types (e.g., "CPU", "GPU", "TPU"): It takes into account the backend/platform types that are involved in the transfer, expecting the user to apply expert knowledge on which backend/platform types are considered costly when used for transfers.

3. Focusing on involved communication channels (e.g., "memcpy", "pcie", "host network", "accelerator interconnect": This instead makes JAX knowledgeable on which underlying communication channels of transfers can be expensive. JAX needs to determine which level it can draw a line, or it may attempt to let the user choose a certain transfer type. Since low-level runtimes typically do not explicitly reveal what communication channels are used, JAX often has to guess communication channel types being used by the low-level runtimes.

The current transfer guard is close to Option 3 with a single threshold that looks at memcpy/relayouting.

As for this feature request to extend the transfer guard API, potential candidates are
(a) Option 2 with multiple backend/paltform type support (proposal in https://github.com/google/jax/issues/16602#issuecomment-1625413621).
(b) Option 2 with XLA:CPU being ignored for all users as a special case.
(c) Option 3 with multiple channel type support.
(d) Option 3 with the single threshold that is raised to ignore local memcpy/relayouting for all users.

Is this list complete? I think we should also hear from JAX core runtime devs.

**Comment by lucasb-eyer** [comment]
Just stumbled upon this too and agree with Alex.

The guard implementation (and most of Jax actually) already makes very clear difference between the concept of "host" and "device": https://github.com/google/jax/blob/7e1278c0400eed18a57b6402f9308c023493f942/jax/_src/config.py#L1218-L1221

With that in mind, I think special-casing "host to host transfer" (ie memcpy/XLA:CPU) as in your (b) suggestion seems to be the best trade-off between practical usefulness, work required, and simplicity.
